### PR TITLE
feat: parse bundler files via ruby parser (WIP)

### LINKED ISSUE #2983: Bundler extract and update using Ruby
The complexity of Gemfile/gemspec probably means we should use Ruby to parse it - perhaps Bundler itself. Here are the complexities:

- Bundler allows users to define multiple groups per dependency. If we translate this to `depType`, it means we need to convert that field into a `depTypes` array field first (for all managers), or support both depType and depTypes
- We need a way to know where in the original file to patch once we have an update. e.g. let's say rails@5.3.0 is released, how do we know exactly where in the file to update it? Referencing line number is unlikely to be a solution. If we know the depTypes (i.e. groups list) and the depName, is that enough to locate it in the file? Are there valid reasons to list a dependency more than once in a Gemfile? Are there valid reasons to list the same group combination twice?

e.g.

```
group :test do
  gem 'guard-rspec', '0.5.5'
end

group :test do
  gem 'rack-google_analytics'
end
```

### PR DESCRIPTION
**Pull request by sleekybadger**
Closes #2983 

Hi. This should help with parsing Gemfiles. It works pretty well even for really complex cases and fetches all meta information that is passed via block or `gem` attributes.

```ruby
ruby '2.5.3'

source 'https://rubygems.org'

group :test do
  gem 'minitest-bisect'
  gem 'minitest-retry', '>= 1.0.0', '< 5', group: :development

  platforms :mri do
    gem 'byebug', source: 'https://external.com'
    gem 'stackprof'
  end

  source 'https://another-external.com' do
    gem 'benchmark-ips'
  end
end
```

```console
{
  "deps": [
    {
      "currentValue": null,
      "depName": "minitest-bisect",
      "depTypes": [
        "test"
      ],
      "lineNumber": [
        6
      ],
      "platforms": [],
      "registryUrls": [
        "https://rubygems.org/"
      ],
      "skipReason": "no-version"
    },
    {
      "currentValue": "< 5, >= 1.0.0",
      "depName": "minitest-retry",
      "depTypes": [
        "test",
        "development"
      ],
      "lineNumber": [
        7
      ],
      "platforms": [],
      "purl": "pkg:rubygems/minitest-retry",
      "registryUrls": [
        "https://rubygems.org/"
      ],
    },
    {
      "currentValue": null,
      "depName": "byebug",
      "depTypes": [
        "test"
      ],
      "lineNumber": [
        10
      ],
      "platforms": [
        "mri"
      ],
      "registryUrls": [
        "https://external.com/"
      ],
      "skipReason": "no-version"
    },
    {
      "currentValue": null,
      "depName": "stackprof",
      "depTypes": [
        "test"
      ],
      "lineNumber": [
        11
      ],
      "platforms": [
        "mri"
      ],
      "registryUrls": [
        "https://rubygems.org/"
      ],
      "skipReason": "no-version"
    },
    {
      "currentValue": null,
      "depName": "benchmark-ips"
      "depTypes": [
        "test"
      ],
      "lineNumber": [
        15
      ],
      "platforms": [],
      "registryUrls": [
        "https://another-external.com/"
      ],
      "skipReason": "no-version"
    },
  ],
  "registryUrls": [
    "https://rubygems.org/",
    "https://external.com/",
    "https://another-external.com/"
  ],
  "rubyVersion": "2.5.3"
}
```

The only thing that bothers me it is conditions inside Gemfile. For example from rails Gemfile:

```ruby
platforms :jruby do
  if ENV["AR_JDBC"]
    gem "activerecord-jdbcsqlite3-adapter", github: "jruby/activerecord-jdbc-adapter", branch: "master"
    group :db do
      gem "activerecord-jdbcmysql-adapter", github: "jruby/activerecord-jdbc-adapter", branch: "master"
      gem "activerecord-jdbcpostgresql-adapter", github: "jruby/activerecord-jdbc-adapter", branch: "master"
    end
  else
    gem "activerecord-jdbcsqlite3-adapter", ">= 1.3.0"
    group :db do
      gem "activerecord-jdbcmysql-adapter", ">= 1.3.0"
      gem "activerecord-jdbcpostgresql-adapter", ">= 1.3.0"
    end
  end
end
```

We can't access `activerecord-jdbcsqlite3-adapter`, `activerecord-jdbcmysql-adapter`, `activerecord-jdbcpostgresql-adapter` gems cause they are inside condition which is controlled via `AR_JDBC` environment variable. The only thing that I was able to come up with is to wrap ENV object and run parsing phaze two times. First time all environment variables handled as true, second time as false then merge results and remove duplicates, but this solution doesn't handle cases where condition will be controlled not via boolean type, for example:

```console
if ENV['RAILS_ENV'] == 'development'
  gem 'pry'
  gem 'web-console'
end
```

### CODE REVIEW

**Comment by rarkins** [comment]
Funny thing is that a "dumb" parser like I did in JS doesn't care about the IF statements and goes right past them to collect every gem.

You could parse twice like this maybe:

- Replace every 'if x' statement with 'if true or x'
- Replace every 'if x' statement with 'if false or x'

But even then you would miss out if there's more than just if/else. e.g. `if X, else if Y, else ...`

**Comment by rarkins** [comment]
This does sound like a difficult problem though. We definitely want every single gem line no matter how many conditional lines there are

**Comment by rarkins** [comment]
@sleekybadger I've made some updates to the JS-based parsing to extract both the Ruby and Bundler versions/constraints if present, and to pick the right Docker image and install the exact Bundler version before updating `Gemfile.lock`. Your PR will also need some brief updates to match. Ruby and Bundler versions should be put into `compatibility.ruby` and `compatibility.bundler` and you can check the new tests cases for an example.

I also fixed the update function in this PR so that it handles multiple line numbers and updates all.

Before merging this PR we'll need to work out how to run the parsing (extract) via Docker too.

**Comment by sleekybadger** [comment]
@rarkins updated

I successfully run it locally against of some my Ruby repos but didn't have a lot of time to verify the correctness of suggested dependencies, so it would be awesome if you can recheck this.

Also, I would like to refactor `bundler/artifacts.js` and `bundler/update.js` files if you don't mind, but I think this can be done in separate PR 😉

**Comment by rarkins** [comment]
@sleekybadger could you make sure that we have a test to confirm we parse multi-line dependencies like this?

```
gem 'xxxx',
    git: 'xxxx',
    branch: 'feature/app'
```

File: lib/manager/bundler/extract.js [hunk 1]
```diff
@@ -1,159 +1,155 @@
...
+const createLogger = level => curry(logger[level], 2);
+
+const createParser = type => async path => {
+  const shell = true;
+  const cwd = __dirname;
+  const bin = './parser/parse';
+
+  const args = getCmdArgs([type, resolve(path)]);
+  const env = getCmdEnv(global.trustLevel);
+  const { stdout } = await exec(`${bin} ${args}`, { cwd, shell, env });
```
Comment (rarkins) [inline lib/manager/bundler/extract.js]: TODO: if binarySource === 'docker' then call out to Docker rather than the code directly

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by sleekybadger** [comment]
Sure)

**Comment by rarkins** [comment]
TODO:
- [ ] Support binarySource=docker
- [ ] Extract Ruby version (as is done in master)
- [ ] Extract Bundler version (as is done in master)

**Comment by Scritik** [comment]
Any news? :)

**Pull request closed without merging**